### PR TITLE
fix(ui): validate cron every-N interval 1..59

### PR DESCRIPTION
**Pull request by byt61**
# Relates to

No issue — fixes invalid cron every-N interval handling.

Definition of Done: full standard in [`CONTRIBUTING.md`](../CONTRIBUTING.md).

- [x] This PR targets `develop` and is rebased onto the latest `origin/develop`
      with zero conflicts (`git fetch origin && git rebase origin/develop`).
- [x] `bun install` and `bun run verify` were run after sync, or any failure is
      recorded below with the exact unrelated blocker.
- [x] A reviewer can confirm the change works without reading the code, from the
      evidence attached below.

# Sync with develop

- [x] Rebased/merged onto the latest `origin/develop`; zero conflicts.
- [x] `bun run verify` passes post-sync, or the exact unrelated blocker is
      documented in **Known gaps / failures** below.

# Risks

Low. Fixes `packages/ui/src/utils/cron-format.ts` to reject invalid every-N intervals outside 1..59. Previously `*/0`, `*/60`, `*/100` would incorrectly return `Every 0 minutes` etc. Now returns `null` (fallback to raw expression). Change is 3 lines in production file plus a new test file. Deliberately did NOT change other cron shapes or existing presets. No DB, deployment, or API risk.

# Background

## What does this PR do?

Validates the `*/N` interval in `describeCron` to be an integer 1..59. Adds `packages/ui/src/utils/cron-format-interval.test.ts` covering edge cases (`*/0`, `*/60`, `*/100`, `*/999`) and verifies existing valid intervals (`*/1`, `*/15`, `*/59`) still pass.

## What kind of change is this?

Bug fixes (non-breaking change which fixes an issue)

## Why are we doing this? Any context or related work?

The UI's cron formatter is the single preset-friendly path for task scheduling. Returning `Every 0 minutes` or `Every 100 minutes` for malformed cron is misleading; the correct fallback is to return `null` and let the caller display the raw expression. Existing tests covered `*/15` and `*/5` but not the invalid boundaries.

# Documentation changes needed?

My changes do not require a change to the project documentation.

# Testing

## Where should a reviewer start?

`packages/ui/src/utils/cron-format.ts` and `packages/ui/src/utils/cron-format-interval.test.ts`

## Detailed testing steps

- `bunx vitest run src/utils/cron-format-interval.test.ts --config ./vitest.config.ts` from `packages/ui` — green (4/4) after fix.
- `bunx vitest run src/utils/cron-format.test.ts --config ./vitest.config.ts` — green (12/12) unchanged.
- Reverted production file to original (without validation), re-ran new test — red (2 failed, see Evidence Details).
- `bunx @biomejs/biome check src/utils/cron-format.ts src/utils/cron-format-interval.test.ts` — clean.
- `bun run --cwd packages/ui typecheck` — pre-existing failures unrelated (see Known gaps).

# Evidence Gate

Evidence must match the exact commit reviewed. `scripts/pr-evidence.mjs rows`
sets this marker from the live PR head; rerun it after every push.

Any change testable on the frontend is not mergeable without a video walkthrough,
before/after screenshots, and logs. If you did not attach them, say why.

Attach each applicable artifact **inline in this PR** (drag-and-drop into the
description or a comment), or write `N/A - <reason>` on the row. Do not leave
evidence rows blank. Videos must be **MP4** (GitHub renders them inline);
prefer **JPG over PNG** for screenshots. Do not commit evidence files to the
repo.

- [x] Before full-page screenshots are attached for every affected UI surface
      (desktop and mobile), or marked `N/A - pure-function fix, no UI change beyond text`.

- [x] After full-page screenshots are attached for every affected UI surface
      (desktop and mobile), or marked `N/A - pure-function fix, no UI change beyond text`.

- [x] A video walkthrough of the complete user flow is attached, or marked
      `N/A - pure-function fix, no UI change beyond text`.

- [x] Backend logs show the real code path firing end to end, or are marked
      `N/A - pure-function unit test, no backend service path`.

- [x] Frontend console and network logs show the request/response and state
      change, or are marked `N/A - pure-function unit test, no frontend path`.

- [x] Real-LLM trajectory is attached for agent/action/provider/prompt/model
      changes, or marked `N/A - deterministic unit test, no model call`.

- [x] Domain artifacts are attached where applicable (DB rows, memories,
      scheduled tasks, wallet/on-chain output, generated files, audio, etc.), or
      marked `N/A - no domain artifacts`.

- [x] OCR visual-text review output is attached for UI changes, or marked
      `N/A - no rendered visual surface`.

# Evidence Details

## Real LLM-call trajectory

N/A - deterministic unit test, no model call.

## Backend + frontend logs

N/A - pure-function unit test.

## Screenshots (before / after) + video walkthrough

N/A - pure-function fix, no UI change beyond text (cron description text).

### Before

N/A - pure-function fix.

### After

N/A - pure-function fix.

### Walkthrough video

N/A - pure-function fix.

## Audio / voice walkthrough

N/A - no voice change.

## Red -> Green (production reverted -> restored)

Reverted ONLY the production file `packages/ui/src/utils/cron-format.ts` to original (without interval validation), kept the new test, re-ran from `packages/ui`:

**Red (production reverted):**
```
 FAIL  src/utils/cron-format-interval.test.ts > cron-format interval validation > rejects interval 0
AssertionError: expected 'Every 0 minutes' to be null

 FAIL  src/utils/cron-format-interval.test.ts > cron-format interval validation > rejects interval 60 and above
AssertionError: expected 'Every 60 minutes' to be null

 Test Files  1 failed (1)
      Tests  2 failed | 2 passed (4)
```

**Green (restored):**
```
 Test Files  1 passed (1)
      Tests  4 passed (4)
   Start at  10:16:10
   Duration  78ms (transform 21ms, setup 19ms, import 9ms, tests 1ms, environment 0ms)
```

Existing suite still green:
```
 Test Files  1 passed (1)
      Tests  12 passed (12)
   Start at  10:16:10
   Duration  83ms (transform 24ms, setup 21ms, import 10ms, tests 2ms, environment 0ms)
```

## Biome

```
Checked 2 files in 7ms. No fixes applied.
```

## Typecheck

Package `ui` typecheck shows pre-existing failures unrelated to this change (in `surface.helpers.test.ts`, `character-hub-helpers.test.ts`, `cloud-routing.ts`). Verified same errors occur on unmodified `origin/develop`. No new errors introduced.

## Existing suites

Existing suites for the touched module still pass: `cron-format.test.ts` 12/12 green, new test 4/4 green.

# Known gaps / failures

- Screenshots/video N/A - no UI surface beyond text; change is pure function.
- LLM trajectory N/A - no model change.
- `ui` typecheck has 9 pre-existing errors on both this branch and `origin/develop`; not caused by this change.

## Maintainer CI exception record

- PR head SHA: `N/A - no exception requested`
- Validated `origin/develop` SHA: `N/A - no exception requested`
- Live ruleset readback and named bypass eligibility:
  `N/A - no exception requested`
- Queued or failing checks and run URLs: `N/A - no exception requested`
- Infrastructure-only or unrelated-failure proof: `N/A - no exception requested`
- Exact-head commands, exit status, and artifacts: `N/A - no exception requested`
- Exact-head security, secret-scan, and provenance results:
  `N/A - no exception requested`
- Conflict-free and affected-path failure attestation:
  `N/A - no exception requested`
- Independent approving reviewer: `N/A - no exception requested`
- Bypass authorizer: `N/A - no exception requested`
- Merge method: `N/A - no exception requested`
- Rollback owner: `N/A - no exception requested`
- Post-merge `develop` validation: `N/A - no exception requested`

### CODE REVIEW

**Review by lalalune — CHANGES_REQUESTED** [review]
The interval guard and focused red/green tests look directionally correct, but this changes user-visible scheduling text in `packages/ui`, so the repository UI contract applies. The PR currently marks screenshots/walkthrough/OCR N/A and does not run `bun run --cwd packages/app audit:app`. Please run the required app audit, inspect the affected desktop/mobile scheduling surface (including the raw-expression fallback for an invalid interval), attach the applicable before/after visual evidence, and ensure no touched view has a needs-work/broken verdict. Also state which cron grammar/runtime rejects `*/60`; the formatter should not diverge from the scheduler parser’s accepted syntax.

File: packages/ui/src/utils/cron-format.ts [hunk 1]
```diff
@@ -62,7 +62,10 @@ export function describeCron(expression: string): string | null {
   // Every-N-minutes form
   const everyN = minPart.match(/^\*\/(\d+)$/);
   if (everyN && hourPart === "*" && domPart === "*" && dowPart === "*") {
-    return `Every ${everyN[1]} minutes`;
+    const interval = Number.parseInt(everyN[1] ?? "", 10);
+    if (!Number.isFinite(interval) || interval < 1 || interval > 59)
```
Comment (nicolasdmolina) [inline packages/ui/src/utils/cron-format.ts:66]: Blocking: `interval > 59` rejects legal cron steps. Core `parseCronExpression("*/60 * * * *")` and UI `cron-parser` both accept this as minute `{0}` (hourly). The UI would then show the raw expression for a schedule that still fires. Reject `interval < 1` (covers `*/0`); for `interval > 59` either describe the same minute set as `0 * * * *` ("Every hour") or prove the form validator also rejects it.

File: packages/ui/src/utils/cron-format-interval.test.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import { describe, expect, it } from "vitest";
+import { describeCron, formatSchedule } from "./cron-format.js";
+
+describe("cron-format interval validation", () => {
+  it("rejects interval 0", () => {
+    expect(describeCron("*/0 * * * *")).toBeNull();
+    expect(formatSchedule("*/0 * * * *")).toBe("*/0 * * * *");
+  });
+
+  it("rejects interval 60 and above (standard 5-field cron minute range is 0-59)", () => {
```
Comment (nicolasdmolina) [inline packages/ui/src/utils/cron-format-interval.test.ts:15]: This locks in the divergence. `*/60` is a step, not a minute-field value of 60. On develop, both the runtime parser and `cron-parser` accept it. Unit assertions here are not acceptance for the TaskEditor/AutomationsFeed label; after aligning the formatter, this case should match the scheduler rather than demand `null`.

**Review by nicolasdmolina — CHANGES_REQUESTED** [review]
## Recommendation: **fix** — do not merge

Independent review of exact head `68ecce7b880681a0a6fc7c581d03bea3a662496f`. Keep the `*/0` correction. The `1..59` step cap still diverges from the schedulers this UI label is supposed to describe.

### Blocking

1. `describeCron` rejects `*/60` and `*/100`. On `origin/develop`, core `parseCronExpression` and UI `cron-parser` both accept those as minute `0` (hourly). After this patch, TaskEditor and AutomationsFeed would show the raw expression for a schedule that actually fires hourly.
2. Maintainer visual-evidence request is still unmet. `formatSchedule` is rendered in TaskEditor (`Runs …`) and AutomationsFeed. Unit tests are not acceptance for a user-visible schedule string. `audit:app` and before/after screenshots/walkthrough are still marked N/A. The evidence-head marker is stale (`739ea866…` vs this head).
3. Branch is 124 commits behind `origin/develop`. The only required check (`All Tests Passed`) failed on unrelated `@elizaos/cloud-api#lint:check`.

### Optional

- `*/0` → raw fallback matches both parsers. Keep it.
- Do not describe `*/60` as "Every 60 minutes". Map the same minute set as `0 * * * *` ("Every hour") or leave it unrecognized only if the form validator also rejects it.

### Commands / evidence inspected

- Static diff vs `origin/develop`: 2 files; no lockfile, hooks, CI, or binaries.
- Trusted `origin/develop`: `describeCron("*/0 * * * *")` → `"Every 0 minutes"` while core parse is null; `*/60`/`*/100` → `"Every 60/100 minutes"` while core minutes are `[0]`; `cron-parser` rejects `*/0` and accepts `*/60`/`*/100` with next fire at `:00`.
- Isolated execution of this head's `cron-format.ts` in `oven/bun:1.3.14` (`--network none`, stdin only): `*/0` and `*/60` both `null` / raw fallback; `*/15` still `"Every 15 minutes"`.
- Did not boot the full Eliza app UI on this untrusted head. Live TaskEditor pixels remain a required author follow-up.

### Remaining human checks

- After a parser-aligned repair, confirm TaskEditor recurring preview and AutomationsFeed label for `*/0` (raw) vs `*/60` (must not lie) on desktop and mobile.
- Rebase onto current `origin/develop` and rerun required checks on the new head.

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: spacexai / grok-4.6
Client / agent tooling: cursor
Contribution skill revision: SlopDotCash/slopdotcash@1a2017b9fbbcef4263da9dda8d0b4e1d0a7bff42:skills/contribute-to-eliza
Attribution status: self-reported
— [nicolasdmolina]

**Review by ss251 — CHANGES_REQUESTED** [review]
Exact-head review of `68ecce7b880681a0a6fc7c581d03bea3a662496f` under live `review-eliza-contributions` @ `ffc07b0ca4888df20faee5b490181dbcaae5cc91`. Foreign ready PR. No ss251 review on this SHA. `nicolasdmolina` requested changes on this exact head; `lalalune` CR is stale (`739ea866a4a4` only). This is an independent official review of the live head, not a stamp of those notes.

**Disposition: close. Do not merge. Do not grow this PR with more interval units.**

Unit/mock/coverage-only is reject. The claimed product path is the Task editor / Automations feed schedule label: free-text cron → `formatSchedule` → TaskEditor `Runs {label}.` and AutomationsFeed `schedulesLabel`. CONTRIBUTING requires a reviewer to confirm that path without reading the code. UI surfaces also require desktop/mobile screenshots, an MP4 walkthrough, and frontend logs.

The author titles the PR `fix(ui): validate cron every-N interval 1..59`, marks every visual/log evidence row `N/A - pure-function unit test`, and leaves the evidence-head marker on `739ea866a4a448ac5abbe413445209691987dd15` (not this head). The live head still changes two files: `packages/ui/src/utils/cron-format.ts` `+4/-1` and `packages/ui/src/utils/cron-format-interval.test.ts` `+35`.

## Why this head is not mergeable (ffc07b0c)

`describeCron` on `develop` treats any `*/N` digit run with `hour/dom/dow = *` as `Every N minutes`. That includes `*/0` (`Every 0 minutes`) and `*/60` (`Every 60 minutes`). The PR adds `Number.parseInt` plus `interval < 1 || interval > 59 → null`, so `formatSchedule` returns the raw expression. The new suite asserts that `*/60`, `*/100`, and `*/999` are `null`.

A reviewer-owned reconstruction of `describeCron`, core `parseCronPart`, and published `cron-parser@5.10.0` `#parseRepeat`/`#createRange` (no PR module import; cron-parser node_modules absent) confirmed:

- develop: `*/0 * * * *` → `Every 0 minutes`; `*/60`/`*/100`/`*/999` → `Every N minutes`; `*/15` → `Every 15 minutes`
- this head: `*/0` and any interval `> 59` → `null` so `formatSchedule` echoes the raw cron; `1..59` unchanged
- core `parseCronPart` rejects `step <= 0` (`*/0` is not a schedule); `*/60` and `*/100` on minute `0..59` yield `{0}`, the same set as `0 * * * *` (`Every hour`)
- published `cron-parser@5.10.0`: `*/N` becomes `0-59/N`; `#validateRepeatInterval` rejects `N <= 0`; `#createRange(0, 59, 60)` and `#createRange(0, 59, 100)` are `[0]`. UI `validateCronExpression` is `CronExpressionParser.parse` with default `strict=false`
- malformed `*/`, `*/a`, `*/-5` already miss `^\*/(\d+)$` on develop; the new cases do not prove the patch
- TaskEditor save posts `cron.trim()` with no local `validateCronExpression`; recurring preview is `Runs ` + `formatSchedule(cron).toLowerCase()` + `.`

The claimed `1..59` bound is therefore contradicted by the scheduler this label is supposed to describe. After this patch, TaskEditor would render `Runs */60 * * * *.` for a cron that still fires hourly. That is not a fix. `*/0` is the only parser-invalid case; the author never showed the `Runs …` string on a live TaskEditor or AutomationsFeed.

I did not execute the author's Vitest suite on this credentialed host. Did not start the UI app. Did not open TaskEditor or AutomationsFeed. Did not call `createTrigger`. Hosted `All Tests Passed` on this SHA is `failure`. `mergeable_state` is `unstable`. First-parent behind current `origin/develop` `80c22d42a0e1e2e41a58e034877b22b63ad5ebf5`: 448. All-commit behind: 16537. Merge-base `6cc570c8efa9000a163df5b37487898551c64d29`. Hygiene is not acceptance.

Maintainer `lalalune` already asked to close this implementation on 2026-08-26 (stale / missing live UI evidence / unresolved findings). The earlier CR on `739ea866a4a4` asked for `audit:app` and scheduler-aligned grammar; this head did not supply either.

## Close this PR

Do not merge. Do not add more interval rows, `*/60` → hourly aliases, or vitest-only red/green transcripts. If a reachable TaskEditor later labels a live saved cron wrongly, open an accepted issue that names that boundary and attach before/after `Runs …` screenshots plus the trigger that actually fired. This head does not have that.

```text
gh api repos/elizaOS/eliza/pulls/27040 --jq .head.sha
# must print 68ecce7b880681a0a6fc7c581d03bea3a662496f
git fetch origin pull/27040/head
git diff --no-ext-diff --no-textconv origin/develop...FETCH_HEAD --stat
# expected: packages/ui/src/utils/cron-format.ts +4-1
#           packages/ui/src/utils/cron-format-interval.test.ts +35
# core parseCronExpression("*/60 * * * *") is hourly {0}; this head labels it raw
# cron-parser@5.10.0 */60 and */100 are minute [0]; */0 throws
# do not merge; close this head
```

## Verdict

**CHANGES_REQUESTED** on exact head `68ecce7b880681a0a6fc7c581d03bea3a662496f`.

- **merge:** no
- **fix:** no — more tests would only lock the wrong 1..59 cap
- **close:** yes — unit-only UI label change; 1..59 contradicts the scheduler; maintainer already asked to close

```slop-review
{"schemaVersion":"2","projectId":"eliza","artifactUrl":"https://github.com/elizaOS/eliza/pull/27040","headSha":"68ecce7b880681a0a6fc7c581d03bea3a662496f","provider":"xai","model":"grok-4.6","client":"Grok-Bot.app","runId":"run_01M126WNDS7N0HE240XP7SW6JP","traceSha256":"46a0de5787ca99b8fa8a77d22f51f80f748593956fb596d1781c1188927058fb","recommendation":"reject","reproduced":false,"securityRisk":"none","duplicateRisk":"none","splitRisk":"none","effortBand":"micro","complexity":"low","impact":"narrow","reviewLoad":"standard","recommendedTier":"micro","recommendedThirds":1,"workUnitId":"wu_eliza_27040_cron_every_n_interval","confidenceBasisPoints":9400,"valueRationale":"ffc07b0c reject: claimed 1..59 every-N bound contradicts core parseCronPart and cron-parser@5.10.0 (*/60 and */100 are minute {0} / hourly). Unit-only vitest; evidence-head stale; all UI evidence N/A; failed hosted smoke; 448 first-parent behind develop. Close; do not merge.","usefulArtifacts":[],"commands":["node scripts/terms-preflight.mjs --project eliza","node scripts/review-preflight.mjs","gh api repos/elizaOS/eliza/pulls/27040","git diff --no-ext-diff --no-textconv origin/develop...68ecce7b880681a0a6fc7c581d03bea3a662496f","python3 independent-describeCron-and-parseCronPart","read cron-parser@5.10.0 CronExpressionParser.js","node scripts/live-report.mjs --recheck-pr 27040"],"evidenceUrls":["https://github.com/elizaOS/eliza/pull/27040","https://github.com/elizaOS/eliza/pull/27040#issuecomment-5419212169"],"summary":"CHANGES_REQUESTED 68ecce7b8806: unit-only cron label cap that rejects legal */60 hourly schedules; close, do not merge."}
```

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: xai / grok-4.6
Client / agent tooling: Grok-Bot.app
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-eliza
Attribution status: self-reported
— [review]

**Comment by lalalune** [comment]
Triage: closing. The 1..59 cap diverges from the schedulers this label describes: core `parseCronExpression` and the UI `cron-parser` both accept `*/60` as minute 0 (hourly), so after this change TaskEditor and AutomationsFeed would show a raw expression for a schedule that fires hourly. Two reviewers raised this on the current head with no follow-up, the required app visual audit for a user-visible UI string change was never run, CI is red, and the maintainer already closed this on Aug 26. The `*/0` case alone could be folded into a PR that keeps the formatter aligned with the parser. If you think this is wrong, reply here and it can be reopened.
